File: package-python/python/perpetual/uplift.py
"""Uplift modelling via the R-Learner approach.

Estimates the Conditional Average Treatment Effect (CATE) using a three-stage
gradient boosting pipeline: outcome model, propensity model, and effect model.
"""

from typing import Any, Dict, Iterable, Optional, Union

import numpy as np
from typing_extensions import Self

from perpetual.perpetual import UpliftBooster as CrateUpliftBooster
from perpetual.utils import convert_input_array, convert_input_frame


class UpliftBooster:
    """R-Learner uplift model for estimating heterogeneous treatment effects.

    Learns the Conditional Average Treatment Effect (CATE)
    ``tau(x) = E[Y | X, W=1] - E[Y | X, W=0]`` using three sequentially
    fitted gradient boosting models: an outcome model, a propensity model,
    and an effect model.
    """

    def __init__(
        self,
        outcome_budget: float = 0.5,
        propensity_budget: float = 0.5,
        effect_budget: float = 0.5,
        num_threads: Optional[int] = None,
        monotone_constraints: Union[Dict[Any, int], None] = None,
        force_children_to_bound_parent: bool = False,
        missing: float = np.nan,
        allow_missing_splits: bool = True,
        create_missing_branch: bool = False,
        terminate_missing_features: Optional[Iterable[Any]] = None,
        missing_node_treatment: str = "None",
        log_iterations: int = 0,
        quantile: Optional[float] = None,
        reset: Optional[bool] = None,
        categorical_features: Union[Iterable[int], Iterable[str], str, None] = "auto",
        timeout: Optional[float] = None,
        iteration_limit: Optional[int] = None,
        memory_limit: Optional[float] = None,
        stopping_rounds: Optional[int] = None,
        max_bin: int = 256,
        max_cat: int = 1000,
        interaction_constraints: Optional[list[list[int]]] = None,
    ):
        """
        Uplift Boosting Machine (R-Learner).

        Estimates the Conditional Average Treatment Effect (CATE): tau(x) = E[Y | X, W=1] - E[Y | X, W=0].

        Parameters
        ----------
        outcome_budget : float, default=0.5
            Fitting budget for the outcome model ``mu(x)``. Higher values allow
            more boosting rounds.
        propensity_budget : float, default=0.5
            Fitting budget for the propensity model ``p(x)``. Higher values allow
            more boosting rounds.
        effect_budget : float, default=0.5
            Fitting budget for the effect model ``tau(x)``. Higher values allow
            more boosting rounds.
        num_threads : int, optional
            Number of threads to use during training and prediction.
        monotone_constraints : dict, optional
            Constraints mapping feature indices/names to -1, 1, or 0.
        force_children_to_bound_parent : bool, default=False
            Whether to restrict children nodes to be within the parent's range.
        missing : float, default=np.nan
            Value to consider as missing data.
        allow_missing_splits : bool, default=True
            Whether to allow splits that separate missing from non-missing values.
        create_missing_branch : bool, default=False
            Whether to create a separate branch for missing values (ternary trees).
        terminate_missing_features : iterable, optional
            Features for which missing branches are always terminated when
            ``create_missing_branch`` is True.
        missing_node_treatment : str, default="None"
            How to handle weights for missing nodes. Options: ``"None"``,
            ``"AssignToParent"``, ``"AverageLeafWeight"``, ``"AverageNodeWeight"``.
        log_iterations : int, default=0
            Logging frequency (every N iterations). 0 disables logging.
        quantile : float, optional
            Target quantile when using ``"QuantileLoss"``.
        reset : bool, optional
            Whether to reset the model or continue training on subsequent fits.
        categorical_features : iterable or str, default="auto"
            Feature indices or names to treat as categorical.
        timeout : float, optional
            Time limit for fitting in seconds.
        iteration_limit : int, optional
            Maximum number of boosting iterations.
        memory_limit : float, optional
            Memory limit for training in GB.
        stopping_rounds : int, optional
            Number of rounds without improvement before stopping.
        max_bin : int, default=256
            Maximum number of bins for feature discretization.
        max_cat : int, default=1000
            Maximum unique categories before a feature is treated as numerical.
        interaction_constraints : list of list of int, optional
            Groups of feature indices allowed to interact.
        """
        self.outcome_budget = outcome_budget
        self.propensity_budget = propensity_budget
        self.effect_budget = effect_budget
        self.num_threads = num_threads
        self.monotone_constraints = monotone_constraints
        self.force_children_to_bound_parent = force_children_to_bound_parent
        self.missing = missing
        self.allow_missing_splits = allow_missing_splits
        self.create_missing_branch = create_missing_branch
        self.terminate_missing_features = terminate_missing_features
        self.missing_node_treatment = missing_node_treatment
        self.log_iterations = log_iterations
        self.quantile = quantile
        self.reset = reset
        self.categorical_features = categorical_features
        self.timeout = timeout
        self.iteration_limit = iteration_limit
        self.memory_limit = memory_limit
        self.stopping_rounds = stopping_rounds
        self.max_bin = max_bin
        self.max_cat = max_cat
        self.interaction_constraints = interaction_constraints

        self.booster = CrateUpliftBooster(
            outcome_budget=outcome_budget,
            propensity_budget=propensity_budget,
            effect_budget=effect_budget,
            max_bin=max_bin,
            num_threads=num_threads,
            monotone_constraints={},
            interaction_constraints=interaction_constraints,
            force_children_to_bound_parent=force_children_to_bound_parent,
            missing=missing,
            allow_missing_splits=allow_missing_splits,
            create_missing_branch=create_missing_branch,
            terminate_missing_features=set(),
            missing_node_treatment=missing_node_treatment,
            log_iterations=log_iterations,
            quantile=quantile,
            reset=reset,
            categorical_features=set(),
            timeout=timeout,
            iteration_limit=iteration_limit,
            memory_limit=memory_limit,
            stopping_rounds=stopping_rounds,
        )

    def fit(self, X, w, y) -> Self:
        """
        Fit the Uplift model.

        Parameters
        ----------
        X : array-like
            Covariates.
        w : array-like
            Treatment indicator (0 or 1).
        y : array-like
            Outcome variable.
        """
        # Feature conversion
        features_, flat_data, rows, cols, categorical_features_, cat_mapping = (
            convert_input_frame(X, "auto", 1000)
        )
        self.feature_names_in_ = features_

        # Treatment conversion
        w_arr = np.array(w)
        if not np.all(np.isin(w_arr, [0, 1])):
            raise ValueError("Treatment indicator 'w' must be binary (0 or 1).")
        w_, _ = convert_input_array(w_arr, "LogLoss")  # Treat as binary

        # Outcome conversion
        y_, _ = convert_input_array(y, "SquaredLoss")  # Treat as regression by default

        self.booster.fit(flat_data, rows, cols, w_, y_)
        return self

    def predict(self, X) -> np.ndarray:
        """
        Predict CATE.

        Parameters
        ----------
        X : array-like
            Covariates.

        Returns
        -------
        cate : ndarray
            Predicted Conditional Average Treatment Effect.
        """
        features_, flat_data, rows, cols, _, _ = convert_input_frame(X, "auto", 1000)

        preds = self.booster.predict(flat_data, rows, cols)
        return preds

    def to_json(self) -> str:
        """Serialize model to JSON string."""
        return self.booster.json_dump()

    @classmethod
    def from_json(cls, json_str: str) -> "UpliftBooster":
        """Deserialize model from JSON string."""
        obj = cls.__new__(cls)
        obj.booster = CrateUpliftBooster.from_json(json_str)
        obj.outcome_budget = None
        obj.propensity_budget = None
        obj.effect_budget = None
        return obj

File: src/causal/uplift.rs
//! Uplift Booster (R-Learner)
//!
//! Estimates the Conditional Average Treatment Effect (CATE) using the
//! R-Learner meta-algorithm backed by gradient boosting.
use crate::booster::config::MissingNodeTreatment;
use crate::booster::core::PerpetualBooster;
use crate::causal::objective::RLearnerObjective;
use crate::constraints::ConstraintMap;
use crate::data::Matrix;
use crate::errors::PerpetualError;
use crate::objective_functions::objective::Objective;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

/// Uplift Booster using the R-Learner Meta-Algorithm.
///
/// Estimates the Conditional Average Treatment Effect (CATE): $\tau(x) = E[Y | X, W=1] - E[Y | X, W=0]$.
///
/// # R-Learner Algorithm
/// 1. **Outcome Model**: $\mu(x) = E[Y|X]$.
/// 2. **Propensity Model**: $p(x) = E[W|X] = P(W=1|X)$.
/// 3. **Effect Model**: Minimizes R-Loss to find $\tau(x)$.
///    $$ L = ((Y - \mu(x)) - \tau(x)(W - p(x)))^2 $$
#[derive(Serialize, Deserialize)]
pub struct UpliftBooster {
    /// Nuisance outcome model $\mu(x) = E[Y|X]$.
    pub outcome_model: PerpetualBooster,
    /// Nuisance propensity model $p(x) = P(W=1|X)$.
    pub propensity_model: PerpetualBooster,
    /// Effect model that learns $\tau(x)$ by minimizing R-Loss.
    pub effect_model: PerpetualBooster,
}

impl UpliftBooster {
    /// Create a new `UpliftBooster` instance using the R-Learner meta-algorithm.
    ///
    /// # Arguments
    ///
    /// * `outcome_budget` - Budget for the outcome model mu(x).
    /// * `propensity_budget` - Budget for the propensity model p(x).
    /// * `effect_budget` - Budget for the effect model tau(x).
    /// * `max_bin` - Maximum number of bins for feature discretization.
    /// * `num_threads` - Optional number of threads.
    /// * `monotone_constraints` - Optional monotonicity constraints.
    /// * `interaction_constraints` - Optional interaction constraints.
    /// * `force_children_to_bound_parent` - Whether to bound child predictions by parent's range.
    /// * `missing` - Missing value representation.
    /// * `allow_missing_splits` - Whether to allow splits isolating missing values.
    /// * `create_missing_branch` - Whether to create explicit missing branches (ternary trees).
    /// * `terminate_missing_features` - Features where missing branches should terminate early.
    /// * `missing_node_treatment` - Strategy for handling missing value splits.
    /// * `log_iterations` - Logging frequency.
    /// * `seed` - Random seed.
    /// * `quantile` - Target quantile for quantile regression.
    /// * `reset` - Whether to reset or continue training on fit.
    /// * `categorical_features` - Features to treat as categorical.
    /// * `timeout` - Hard limit for fitting time in seconds.
    /// * `iteration_limit` - Hard limit for number of iterations.
    /// * `memory_limit` - Memory limit in GB.
    /// * `stopping_rounds` - Number of rounds for early stopping.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        outcome_budget: f32,
        propensity_budget: f32,
        effect_budget: f32,
        max_bin: u16,
        num_threads: Option<usize>,
        monotone_constraints: Option<ConstraintMap>,
        interaction_constraints: Option<Vec<Vec<usize>>>,
        force_children_to_bound_parent: bool,
        missing: f64,
        allow_missing_splits: bool,
        create_missing_branch: bool,
        terminate_missing_features: HashSet<usize>,
        missing_node_treatment: MissingNodeTreatment,
        log_iterations: usize,
        seed: u64,
        quantile: Option<f64>,
        reset: Option<bool>,
        categorical_features: Option<HashSet<usize>>,
        timeout: Option<f32>,
        iteration_limit: Option<usize>,
        memory_limit: Option<f32>,
        stopping_rounds: Option<usize>,
    ) -> Result<Self, PerpetualError> {
        // Base configurations
        let base_params = |budget: f32, objective: Objective| -> Result<PerpetualBooster, PerpetualError> {
            PerpetualBooster::new(
                objective,
                budget,
                f64::NAN,
                max_bin,
                num_threads,
                monotone_constraints.clone(),
                interaction_constraints.clone(),
                force_children_to_bound_parent,
                missing,
                allow_missing_splits,
                create_missing_branch,
                terminate_missing_features.clone(),
                missing_node_treatment,
                log_iterations,
                seed,
                quantile,
                reset,
                categorical_features.clone(),
                timeout,
                iteration_limit,
                memory_limit,
                stopping_rounds,
            )
        };

        let outcome_model = base_params(outcome_budget, Objective::SquaredLoss)?; // Assuming continuous Y for now
        let propensity_model = base_params(propensity_budget, Objective::LogLoss)?; // Treatment is binary

        // The effect model uses the custom R-Learner objective, but we can't initialize it
        // without data (the residuals).
        // So we initialize it with a dummy Objective::SquaredLoss and swap it during fit?
        // OR, since Objective is an enum, we can't easily swap to a Custom one dynamically without wrapping.
        // The `PerpetualBooster` takes an `Objective` in its config.

        // Actually, we can initialize it with SquaredLoss as a placeholder,
        // and then in `fit`, we construct the `RLearnerObjective` and use it.
        // However, `PerpetualBooster`'s config is public.
        let effect_model = base_params(effect_budget, Objective::SquaredLoss)?;

        Ok(UpliftBooster {
            outcome_model,
            propensity_model,
            effect_model,
        })
    }

    /// Fit the Uplift Model.
    ///
    /// * `X`: Covariates
    /// * `w`: Treatment indicator (0 or 1)
    /// * `y`: Outcome
    pub fn fit(&mut self, x: &Matrix<f64>, w: &[f64], y: &[f64]) -> Result<(), PerpetualError> {
        // 1. Fit Outcome Model: Y ~ X
        // R-Learner typically fits mu(x) on all data (ignoring W, or including W?).
        // Nie & Wager suggest cross-fitting, but for a simple implementation:
        // We can just fit Y ~ X. (This is a simplified "nuisance" parameter estimation).
        self.outcome_model.fit(x, y, None, None)?;
        let mu_hat = self.outcome_model.predict(x, true);

        // 2. Fit Propensity Model: W ~ X
        self.propensity_model.fit(x, w, None, None)?;
        // Helper to get probabilities from LogLoss prediction (which are log-odds)
        let log_odds = self.propensity_model.predict(x, true);
        let p_hat: Vec<f64> = log_odds.iter().map(|lo| 1.0 / (1.0 + (-lo).exp())).collect();

        // 3. Prepare R-Learner Objective
        // We need to pass the treatment and the *predicted* nuisance parameters to the objective.
        let r_obj_fn = RLearnerObjective::new(w.to_vec(), mu_hat, p_hat);

        // 4. Update Effect Model's Objective
        // This is the key "Deep Integration" step.
        // We swap the objective of the effect_model to our custom one.
        self.effect_model.cfg.objective = Objective::new_custom(r_obj_fn);

        // 5. Fit Effect Model: Tau ~ X
        // The target 'y' passed to fit is actually used in the loss function context.
        // Our custom loss function uses the 'y' passed here as the actual 'y' from the dataset.
        // It internalizes mu_hat and p_hat.
        self.effect_model.fit(x, y, None, None)?;

        Ok(())
    }

    pub fn predict(&self, x: &Matrix<f64>) -> Vec<f64> {
        self.effect_model.predict(x, true)
    }
}
